web/bigquery/definitions/standard/02_page_views/01_main/07_commit_page_views.sqlx

```diff
@@ -0,0 +1,90 @@
+/*
+   Copyright 2021 Snowplow Analytics Ltd. All rights reserved.
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+       http://www.apache.org/licenses/LICENSE-2.0
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+config {
+  type: "operations",
+  name: "07_commit_page_views",
+  disabled: sp.model_disabled,
+  dependencies: ["06_page_views_metadata"],
+  hermetic: true,
+  hasOutput: false,
+  tags: ["sp_page_views"]
+}
+
+js {
+  function if_derived_pv() {
+    if (sp.skip_derived.page_views === false) {
+      return `
+        -- Commit to production if enabled
+        -- Note: Automigrate hardcoded to false as all columns are to be explicitly defined in model.
+        CALL \`${sp.output_schema}.commit_table\` (
+          '${sp.scratch_schema}',             -- sourceDataset
+          'page_views_this_run${sp.entropy}', -- sourceTable
+          '${sp.output_schema}',              -- targetDataset
+          'page_views${sp.entropy}',          -- targetTable
+          'page_view_id',                     -- joinKey
+          'start_tstamp',                     -- partitionKey
+          FALSE);                             -- automigrate
+      `;
+    } else {
+      return ``;
+    }
+  }
+
+  function if_stage_pv() {
+    if (sp.stage_next.page_views === true) {
+      return `
+        -- Commit staging table if enabled
+        -- Note: Automigrate hardcoded to false as all columns are to be explicitly defined in model.
+        CALL \`${sp.output_schema}.commit_table\` (
+          '${sp.scratch_schema}',             -- sourceDataset
+          'page_views_this_run${sp.entropy}', -- sourceTable
+          '${sp.scratch_schema}',             -- targetDataset
+          'page_views_staged${sp.entropy}',   -- targetTable
+          'page_view_id',                     -- joinKey
+          'start_tstamp',                     -- partitionKey
+          FALSE);                             -- automigrate
+      `;
+    } else {
+      return ``;
+    }
+  }
+}
+
+BEGIN
+
+  ${if_derived_pv()}
```

Comment:
Thank you for the review @bill-warner !
I also share the concerns that you mentioned in your first comment, which are related to the previous two:

> around how the end user interacts with the derived and staging tables

> to include an etl_tstamp on each derived table

> How does the user query the derived pv table?

If a downstream model uses the derived tables, and is defined independently (e.g. its own Dataform project), then it does make sense to consider the derived tables as external to it and add them as its declarations, so they can be `ref()`ed. In that case, a `table` type should consume the `derived` at the time of reference, while for an `incremental` type, it would need to define the incrementalization login itself, which indeed could be made easier with the `etl_tstamp` you suggested. 
On the latter (adding the timestamp), we could certainly discuss about deviating from the parent BQ model, as so far the approach was to not deviate at all at least for v0.1.0.

For the model to be extended with custom modules (same project/schedule), then the user has the ability to define the custom steps as dependent from any action of the model that creates/updates/completes the staged/derived table of interest. In other words, the user can keep on building upon the model the same way (explicitly enforcing dependencies on actions and not implicitly on tables). With Dataform being a declarative tool, it is tricky to retain absolute consistency with the parent model without missing out some of its automated features, at least for this first v0.1.0. 

What do you think?

Reply:
Thanks for the clarification. I agree with it being sufficient to have the derived tables defined as sources within another DF project. 

Regarding building off the staged tables, I think the necessity to declare the dependancies is fine for now. We could explore in the future creating a view per staged table within Dataform, that simply SELECT * from the actual table. Within this view we could define the dependancy to the operation that creates said staged table within the main model. This might simplify things from the users perspective.